Civi/RemoteParticipant/EventSubscriber/MailingListSubscriptionSubscriber.php

```diff
@@ -0,0 +1,180 @@
+
+  private MailingListSubscriptionManager $subscriptionManager;
+
+  /**
+   * @inheritDoc
+   */
+  public static function getSubscribedEvents(): array {
+    return [
+      GetCreateParticipantFormEvent::NAME => ['onGetCreateParticipantForm', -1],
+      GetUpdateParticipantFormEvent::NAME => ['onGetUpdateParticipantForm', -1],
+      ValidateEvent::NAME => ['onValidateEvent', -1],
+      RegistrationEvent::NAME => ['onRegistrationEvent', \CRM_Remoteevent_Registration::AFTER_PARTICIPANT_CREATION],
+      UpdateEvent::NAME => ['onUpdateEvent', \CRM_Remoteevent_RegistrationUpdate::AFTER_APPLY_PARTICIPANT_CHANGES],
+    ];
+  }
+
+  public function __construct(Api4Interface $api4, MailingListSubscriptionManager $subscriptionManager) {
+    $this->api4 = $api4;
+    $this->subscriptionManager = $subscriptionManager;
+  }
+
+  /**
+   * @throws \CRM_Core_Exception
+   */
+  public function onGetCreateParticipantForm(GetCreateParticipantFormEvent $event): void {
+    $this->addFields($event);
+  }
+
+  /**
+   * @throws \CRM_Core_Exception
+   */
+  public function onGetUpdateParticipantForm(GetUpdateParticipantFormEvent $event): void {
+    $this->addFields($event);
+  }
+
+  public function onValidateEvent(ValidateEvent $event): void {
+    /**
+     * @phpstan-var array<int, string> $groupIds
+     *   Mapping of group ID as integer to group ID as string.
+     */
+    $groupIds = $event->getSubmission()['mailing_list_group_ids'] ?? [];
+    /** @phpstan-var list<int> $allowedGroupIds */
+    $allowedGroupIds = $event->getEvent()['event_remote_registration.mailing_list_group_ids'] ?? [];
+    if ([] !== array_diff($groupIds, $allowedGroupIds)) {
+      $l10n = $event->getLocalisation();
+      $event->addValidationError('mailing_list_group_ids', $l10n->ts('Invalid value'));
+    }
+  }
+
+  /**
+   * @throws \CRM_Core_Exception
+   */
+  public function onRegistrationEvent(RegistrationEvent $event): void {
+    $this->subscribeToMailingLists($event);
+  }
+
+  /**
+   * @throws \CRM_Core_Exception
+   */
+  public function onUpdateEvent(UpdateEvent $event): void {
+    $this->subscribeToMailingLists($event);
+  }
+
+  /**
+   * @throws \CRM_Core_Exception
+   */
+  private function addFields(GetParticipantFormEventBase $event): void {
+    $groupIds = $event->getEvent()['event_remote_registration.mailing_list_group_ids'] ?? [];
+    if ([] === $groupIds) {
+      return;
+    }
+
+    $groups = $this->api4->execute('Group', 'get', [
+      'select' => ['id', 'title'],
+      'where' => [
+        ['id', 'IN', $groupIds],
+        ['is_active', '=', TRUE],
+      ],
+      'orderBy' => ['title' => 'ASC'],
+    ])->indexBy('id')
+      ->column('title');
+    if ([] === $groups) {
+      return;
+    }
+
+    $label = $event->getEvent()['event_remote_registration.mailing_list_subscriptions_label'] ?? '';
+    if ('' === $label) {
+      $label = $event->getLocalisation()->ts('I want to subscribe to the following newsletters');
```

Comment:
Let's call it *mailing lists* instead of *newsletters*, as the latter term is not really used in CiviCRM.

Reply:
Which word would you suggest in the German translation? Currently it's *Newsletter*. For the texts in CiviCRM I've chosen *Rundschreiben* which is the word that is used in CiviMail. However, here we have a string shown in the registration form. I think *Rundschreiben* is, even if "more" German, uncommon in this context.

Reply:
*Rundschreiben* is the "official" translation of a CiviCRM *Mailing*, while a *Mailing List* (which is a CiviCRM *Group* of that type) is not being translated at all, transifex translation uses different terms like *E-Mail-Newsletter*, *Verteiler*, *Mailingliste*.

I would suggest calling it a *mailing list* in English and using *E-Mail-Newsletter* in German, although that's an anglicism :shrug:

Reply:
I've changed the strings as discussed.